Discover and check operator variants

amjames opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #76901
* #76900

Operator variants can now be explicitly specified in the OpInfo kwargs.
When the operator name is not the same as the method/function form this
will allow them to be discovered.

The OpInfo is extended to also accept/discover the inplace operator
variant.

Operator and inplace operator variants are exercised in consistency
tests when the sample does not contain any kwargs.

amjames committed:
Discover and check operator variants

Operator variants can now be explicitly specified in the OpInfo kwargs.
When the operator name is not the same as the method/function form this
will allow them to be discovered.

The OpInfo is extended to also accept/discover the inplace operator
variant.

Operator and inplace operator variants are exercised in consistency
tests when the sample does not contain any kwargs.

[ghstack-poisoned]

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -595,6 +595,17 @@ class OpInfo(object):
+    operator_variant: Callable = _NOTHING
+
+    # explicitly specifies the inplace operator variant of the
+    # operator if _NOTHING (default), the inplace operator variant
```
This is confusing to read to read because it seems like it's saying "the inplace operator variant of the operator if _NOTHING..." but it intends to say:

```
# allows the inplace operator variant of this operation to be specified as follows:
  - if _NOTHING (default), then the OpInfo attempts to discover the variant using its name
  - if None, then the OpInfo explicitly specifies it has no associated inplace operator
  - if a Callable, then that callable should be the inplace operator associated with this operation
```

We could make the same change to `operator_variant`, `inplace_variant`, and `method_variant` above for consistency if you also think that's clearer, @amjames

amjames replied:
@mruberry I agree that form of the comment is much clearer. I also may add a note here or below where the `inplace_operator_variant` discovery happens indicating why automatic discovery is disabled if the `inplace_variant` is not defined. 

The `operator` module will attempt to help out by implementing the in-place operator using the out-of-place operator and assigning the result to the lhs. This "works" but does not have proper in-place behavior and will fail tests.

mruberry commented on test/test_ops.py:
```diff
@@ -917,6 +924,10 @@ def _test_inplace_preserve_storage(samples, variants):
                         cloned if isinstance(cloned, torch.Tensor) else cloned[0]
                     )
                     data_ptr = inp_tensor.data_ptr()
+                    if variant in operators and sample.kwargs:
```
Nice skip and comment

mruberry commented on test/test_ops.py:
```diff
@@ -773,16 +773,18 @@ def _case_four_transform(t):
-            inplace,
-        ]
-        variants = [method, inplace]
+        inplace_ops = [inplace, inplace_operator]
```
Nice extension

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -912,12 +932,17 @@ def get_inplace(self):
         """
         return self.operator_variant
 
+    def get_inplace_operator(self):
```
Nice getter modifications

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -9826,6 +9851,8 @@ def generate_std_var_kwargs(t: torch.Tensor, **kwargs):
     BinaryUfuncInfo('bitwise_and',
                     dtypes=integral_types_and(torch.bool),
                     supports_autograd=False,
+                    operator_variant=operator.and_,
```
Cool additions

mruberry commented:
These PRs look great, @amjames! I triggered some additional tests (with the ciflow/all label) just in case. I also made one suggestion on comment readability -- which is not the fault of this PR, but while we're here I thought we might fix it, anyway, if you agree.

Let me know your thoughts and then once these tests pass I think this is good to land!

Lezcano commented:
Are there operations that we were testing manually before and now they are being collected automatically? If so, could you remove the `OpInfos` for those operations in this PR?

amjames commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -817,7 +829,15 @@ def __post_init__(self):
+            self.operator_variant = getattr(operator, self.name, None)
+
+        if self.inplace_operator_variant is _NOTHING:
+            if self.inplace_variant is not None:
```
@mruberry This is the logic I think may warrant additional explanation.

mruberry replied:
Sounds great -- it's hard to have too many comments

amjames committed:
Update on "Discover and check operator variants"


Operator variants can now be explicitly specified in the OpInfo kwargs.
When the operator name is not the same as the method/function form this
will allow them to be discovered.

The OpInfo is extended to also accept/discover the inplace operator
variant.

Operator and inplace operator variants are exercised in consistency
tests when the sample does not contain any kwargs.

[ghstack-poisoned]

Lezcano commented:
Doesn't this PR cover `__rmatmul__`? If so, what would be stopping us from adding the right-looking variants? It'd be so nice to get rid of `__rmatmul__`...

mruberry commented:
> @Lezcano I could not find any `OpInfos` that are redundant due to the new operator capture rules. The only operators for which `OpInfos` are defined right now are the `__r<op>__` variants. This PR does add `OpInfo` coverage for many `__i<op>__` and a few `__<op>__` methods which were missing before.
> 
> I did not add these reverse dunders to be auto-captured as they require additional kwargs to ensure they are not called with a `Tensor` on the LHS. There is currently an open issue (#74616) regarding problems coming from tests where reverse dunders are invoked manually with a `Tensor` on the lhs.

Instead of additional kwargs we might just want to create a sample inputs function that's just for the reverse dunders.

mruberry commented:
Just ping me when this is ready for the next review, @amjames

Lezcano commented:
On the r-looking variants: My concern is that, at the moment, we are writing OpInfos for them manually and that's not great. For example, for non-trivial ones, like the one for `matmul`, the OpInfos for the normal operation and the r-looking one have diverged, and it's highly non-trivial to figure out whether they are equivalent or which one is correct, or whether both are missing things.

But again, perhaps merging the normal and r-looking operations is easier said than done, but yeah.

amjames committed:
Update on "Discover and check operator variants"


Operator variants can now be explicitly specified in the OpInfo kwargs.
When the operator name is not the same as the method/function form this
will allow them to be discovered.

The OpInfo is extended to also accept/discover the inplace operator
variant.

Operator and inplace operator variants are exercised in consistency
tests when the sample does not contain any kwargs.

[ghstack-poisoned]

amjames commented:
@mruberry This is ready for the next review. I am going to explore the idea of adding r-dunders re @Lezcano 's ideas and will add a PR to the stack if that bears any fruit. CI failures appear to be infra related so hopefully they will all be green this time around.

amjames committed:
Update on "Discover and check operator variants"


Operator variants can now be explicitly specified in the OpInfo kwargs.
When the operator name is not the same as the method/function form this
will allow them to be discovered.

The OpInfo is extended to also accept/discover the inplace operator
variant.

Operator and inplace operator variants are exercised in consistency
tests when the sample does not contain any kwargs.

[ghstack-poisoned]

amjames committed:
Update on "Discover and check operator variants"


Operator variants can now be explicitly specified in the OpInfo kwargs.
When the operator name is not the same as the method/function form this
will allow them to be discovered.

The OpInfo is extended to also accept/discover the inplace operator
variant.

Operator and inplace operator variants are exercised in consistency
tests when the sample does not contain any kwargs.

[ghstack-poisoned]

mruberry commented:
Exploring more on the reverse dunders sounds cool but if the simplest solution is to keep them separate and give them their own sample inputs function that also sounds totally fine. Trying to have two sample inputs for already important and complicated operations like matmul might be too unreadable to be worthwhile.

Lezcano commented:
The inputs to rmatmul are those from matmul reversed. That's what led me to suggest automating all this really.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -12109,6 +12150,8 @@ def generate_std_var_kwargs(t: torch.Tensor, **kwargs):
     BinaryUfuncInfo('bitwise_or',
                     ref=np.bitwise_or,
                     dtypes=integral_types_and(torch.bool),
+                    operator_variant=operator.or_,
```
Nice attention to the bitwise operators

mruberry reviewed (approved):
LGTM! We should wait for the tests to finish, but if they're OK you can merge this using `@pytorchbot merge this`

amjames committed:
Update on "Discover and check operator variants"


Operator variants can now be explicitly specified in the OpInfo kwargs.
When the operator name is not the same as the method/function form this
will allow them to be discovered.

The OpInfo is extended to also accept/discover the inplace operator
variant.

Operator and inplace operator variants are exercised in consistency
tests when the sample does not contain any kwargs.

[ghstack-poisoned]

amjames committed:
Update on "Discover and check operator variants"


Operator variants can now be explicitly specified in the OpInfo kwargs.
When the operator name is not the same as the method/function form this
will allow them to be discovered.

The OpInfo is extended to also accept/discover the inplace operator
variant.

Operator and inplace operator variants are exercised in consistency
tests when the sample does not contain any kwargs.

[ghstack-poisoned]

amjames committed:
Update on "Discover and check operator variants"


Operator variants can now be explicitly specified in the OpInfo kwargs.
When the operator name is not the same as the method/function form this
will allow them to be discovered.

The OpInfo is extended to also accept/discover the inplace operator
variant.

Operator and inplace operator variants are exercised in consistency
tests when the sample does not contain any kwargs.

[ghstack-poisoned]

amjames committed:
Update on "Discover and check operator variants"


Operator variants can now be explicitly specified in the OpInfo kwargs.
When the operator name is not the same as the method/function form this
will allow them to be discovered.

The OpInfo is extended to also accept/discover the inplace operator
variant.

Operator and inplace operator variants are exercised in consistency
tests when the sample does not contain any kwargs.

[ghstack-poisoned]

amjames commented:
@pytorchbot merge this
